src/adapt/proto.ts

```diff
@@ -0,0 +1,478 @@
+  REPEATED: FieldDescriptorProto.Label.LABEL_REPEATED,
+  [protos.google.cloud.bigquery.storage.v1.TableFieldSchema.Mode.REQUIRED]:
+    FieldDescriptorProto.Label.LABEL_OPTIONAL,
+  REQUIRED: FieldDescriptorProto.Label.LABEL_REQUIRED,
+  [protos.google.cloud.bigquery.storage.v1.TableFieldSchema.Mode
+    .MODE_UNSPECIFIED]: null,
+  MODE_UNSPECIFIED: null,
+};
+
+// convertStorageSchemaToProto2Descriptor builds a FileDescriptorSet for a given table schema using proto2 syntax.
+export function convertStorageSchemaToProto2Descriptor(
+  schema: TableSchema,
+  scope: string
+): FileDescriptorSet | null {
+  return convertStorageSchemaToFileDescriptorInternal(schema, scope, false);
+}
+
+// convertStorageSchemaToProto3Descriptor builds a FileDescriptorSet for a given table schema using proto3 syntax.
+export function convertStorageSchemaToProto3Descriptor(
+  schema: TableSchema,
+  scope: string
+): FileDescriptorSet | null {
+  return convertStorageSchemaToFileDescriptorInternal(schema, scope, true);
+}
+
+function convertStorageSchemaToFileDescriptorInternal(
+  schema: TableSchema,
+  scope: string,
+  useProto3: boolean
+): FileDescriptorSet {
+  let fNumber = 0;
+  const fields: FieldDescriptorProto[] = [];
+  const deps: Record<string, FileDescriptorProto> = {};
+  for (const field of schema.fields || []) {
+    fNumber += 1;
+    const currentScope = `${scope}_${field.name}`;
+    if (field.type === TableFieldSchema.Type.STRUCT) {
+      const subSchema: TableSchema = {
+        fields: field.fields,
+      };
+      const fd = convertStorageSchemaToFileDescriptorInternal(
+        subSchema,
+        currentScope,
+        useProto3
+      );
+      fd.file?.forEach(f => {
+        if (f.name) {
+          deps[f.name] = f;
+        }
+      });
+      const fdp = convertTableFieldSchemaToFieldDescriptorProto(
+        field,
+        fNumber,
+        currentScope,
+        useProto3
+      );
+      fields.push(fdp);
+    } else {
+      const fdp = convertTableFieldSchemaToFieldDescriptorProto(
+        field,
+        fNumber,
+        currentScope,
+        useProto3
+      );
+      fields.push(fdp);
+    }
+  }
+
+  const dp = new DescriptorProto({
+    name: scope,
+    field: fields,
+  });
+
+  const depsNames: string[] = Object.keys(deps);
+  const syntax = useProto3 ? 'proto3' : 'proto2';
+  const fdp = new FileDescriptorProto({
+    messageType: [dp],
+    name: `${scope}.proto`,
+    syntax,
+    dependency: depsNames,
+  });
+
+  const fds = new FileDescriptorSet({
+    file: [fdp, ...Object.values(deps)],
+  });
+
+  return fds;
+}
+
+// normalizeDescriptor builds a self-contained DescriptorProto suitable for communicating schema
+// information with the BigQuery Storage write API.  It's primarily used for cases where users are
+// interested in sending data using a predefined protocol buffer message.
+export function normalizeDescriptor(fds: FileDescriptorSet): DescriptorProto {
```

Comment:
This method signature seems odd.  A FileDescriptorSet can contain multiple file descriptors, each with multiple messages.  Don't we need to indicate which message becomes the DescriptorProto?  This logic appears to simply pick the first message from the first file in the set, and embed all the others as nested.

Reply:
This is because the method `convertStorageSchemaToProto2Descriptor` is converting the BQ Schema to a `FileDescriptorSet`, as in NodeJS I couldn't find a equivalent to a `protoreflect.MessageDescriptor` that bundle information of the proto message and dependencies. I think for Node, I'll the `convertStorageSchemaToProto2Descriptor` to call `normalizeDescriptor` and return a DescriptorProto. 

I'll think more about that and try to avoid exposing/returning/accepting a FileDescriptorSet on the methods.